File: frontend/src/core/factories/ValidatorFactory.ts
/**
 * Factory for assembling validator chains — centralises node wiring (OCP).
 *
 * Responsibility (SRP): know which validator nodes exist and in what order.
 *     Does not validate anything; does not contain business rules.
 * Depends on: concrete validator classes in modules/tickets/validators/.
 *     This is the ONE place in the FE that imports concrete validator classes (DIP inversion point).
 * Pattern: Factory — decouples chain creation from consumption.
 * SOLID: OCP · SRP · DIP
 *
 * OCP extension (Sprint 4 — CriticalPriorityValidator):
 *   1. Create modules/tickets/validators/CriticalPriorityValidator.ts
 *   2. Add one line in buildTicketChain():
 *        businessRuleV.addValidator(new CriticalPriorityValidator())
 *   3. Nothing else changes.
 */

import { BasicFieldValidator } from '../../modules/tickets/validators/BasicFieldValidator'
import { FileValidator } from '../../modules/tickets/validators/FileValidator'
import { BusinessRuleValidator } from '../../modules/tickets/validators/BusinessRuleValidator'
import type { BaseValidator } from '../base/BaseValidator'

export class ValidatorFactory {
  /**
   * Assemble the ticket-creation validation chain.
   *
   * Chain order (fail-fast left to right):
   *   BasicFieldValidator → FileValidator → BusinessRuleValidator
   *
   * @returns Root node. Caller invokes root.run(data).
   *
   * OCP note: add CriticalPriorityValidator in Sprint 4 by appending one addValidator() call.
   */
  static buildTicketChain(): BaseValidator {
    const basicFieldV      = new BasicFieldValidator()
    const fileV            = new FileValidator()
    const businessRuleV    = new BusinessRuleValidator()

    basicFieldV.addValidator(fileV).addValidator(businessRuleV)

    return basicFieldV
  }
}

File: backend/core/factories/validator_factory.py
"""
Factory for assembling validator chains — centralises node wiring (OCP).

Responsibility (SRP): know which validator nodes exist and in what order to chain them.
Depends on: concrete validator classes in apps/tickets/validators/ — DIP inversion point.
Pattern: Factory.
SOLID: OCP · SRP · DIP
"""

from __future__ import annotations

from core.base.base_validator import BaseValidator


class ValidatorFactory:
    """Creates and wires validator chains for each domain context."""

    @staticmethod
    def build_ticket_chain(ticket_repository) -> BaseValidator:
        """
        Assemble the ticket-creation chain: BasicField → File → Profile → BusinessRule.

        OCP: add CriticalPriorityValidator (Sprint 4) by appending one line:
            business_v.add_validator(CriticalPriorityValidator())
        Nothing else changes.
        """
        from apps.tickets.validators.basic_field_validator import BasicFieldValidator
        from apps.tickets.validators.file_validator import FileValidator
        from apps.tickets.validators.profile_completeness_validator import (
            ProfileCompletenessValidator,
        )
        from apps.tickets.validators.business_rule_validator import BusinessRuleValidator
        from apps.authentication.validators import EmpresaValidator, IdentificationValidator

        basic_field_v = BasicFieldValidator()
        file_v        = FileValidator()
        profile_v     = ProfileCompletenessValidator((
            IdentificationValidator(),
            EmpresaValidator(),
        ))
        business_v    = BusinessRuleValidator(ticket_repository)

        basic_field_v.add_validator(file_v).add_validator(profile_v).add_validator(business_v)
        return basic_field_v
